batch/batch/worker/worker.py

```diff
@@ -1656,32 +1656,14 @@ async def run(self):
 
                 log.info(f'{self}: downloading JAR')
                 with self.step('downloading_jar'):
-                    async with self.worker.jar_download_locks[self.revision]:
-                        local_jar_location = f'/hail-jars/{self.revision}.jar'
-                        if not os.path.isfile(local_jar_location):
-                            self.verify_is_acceptable_query_jar_url(self.jar_url)
-                            temporary_file = tempfile.NamedTemporaryFile(delete=False)
-                            try:
-                                async with await self.worker.fs.open(self.jar_url) as jar_data:
-                                    while True:
-                                        b = await jar_data.read(256 * 1024)
-                                        if not b:
-                                            break
-                                        written = await blocking_to_async(worker.pool, temporary_file.write, b)
-                                        assert written == len(b)
-                                temporary_file.close()
-                                os.rename(temporary_file.name, local_jar_location)
-                            finally:
-                                temporary_file.close()  # close is idempotent
-                                try:
-                                    os.remove(temporary_file.name)
-                                except OSError as err:
-                                    if err.errno != errno.ENOENT:
-                                        raise
+                    local_jar_location = await self.run_until_done_or_deleted(self.download_jar)
 
                 log.info(f'{self}: running jvm process')
                 with self.step('running'):
-                    await self.jvm.execute(local_jar_location, self.scratch, self.user_command_string)
+                    await self.run_until_done_or_deleted(
+                        self.jvm.execute, local_jar_location, self.scratch, self.user_command_string
+                    )
```

Comment:
cc @daniel-goldstein 

---

Just to recap for clarity how the JVMEntryway works: there's a JVM process. That process shares a socket with the worker. The worker-JVMEnryway protocol follows:
1. Worker sends the number of "arguments". [1]
2. Worker sends that number of "arguments" as strings (strings are encoded as length in bytes then data in utf-8)
3. The worker may now optionally send the number 0 to stop execution immediately.
4. The JVMEntryway sends a message (an integer) indicating why it has stopped executing this request. If the worker sent a 0 (requesting immediate stop to execution) we expect to receive the JVM.FINISH_CANCELLED message. The JVM can also stop executing due to a JVMEntryway exception, a user exception, or normal successful exit. We might also receive End of Stream (EOS) which indicates the JVM has died.

---

I think this change can leave the worker in a bad state. For example, if we cancel while `JVM.execute` is half-way through the protocol with the JVM. Unlike, say, an HTTP request, the JVM won't somehow timeout and let the worker start a new session. It's important that the three required steps above are followed. Moreover, we need to tell the JVM to stop processing a user's job early if we want to re-use that JVM.

---

I think we shouldn't change `JVM.execute` to rely on task cancellation. Task cancellation is a blunt instrument and I think it's really easy for us to not handle it correctly. `JVM.execute` instead is explicit about when it can exit early: while it is waiting for the JVM. It's also easy to see that the protocol is followed correctly. We don't have to reason about control jumping from somewhere in the middle of the method into an `except` or `finally` block.

We should carefully inspect any Task with which we plan to use `run_until_done_or_deleted`. That task needs to be written such that cancellation does not leave the worker in a bad state.

`run_until_done_or_deleted` on `borrow_jvm` seems OK. That will let us cancel a job if JVM initialization is hung.

`run_until_done_or_deleted` on `download_jar` is also, I think, OK. The rename is atomic. The finally will cleanup any temporary resources.

---

[1] The "arguments" are: colon-separated classpath, main class name, scratch directory, and then an arbitrary number of arguments which are passed along to the main class's main method.

Reply:
Other than this, I think this PR is excellent and I'm really glad that we can actually measure the timing of writing logs and retrieving output.

I had not intended for those comments to make it into the final PR. I had screwed myself up several times where an exception occurred in `retrieve_and_clear_output` and the logs were lost. Thanks for cleaning up my mess 😅

Reply:
Ah, you know what, this must explain why we had issues with bad images. That `localize_rootfs` method really shouldn't be cancelled once it starts extracting the filesystem.

I think `run_container` is fine. I am less certain about aborting a `mount` or modification of the network namespaces. What's the harm in letting those operations run to completion? Thinking back to explicit opportunities to stop early rather than bluntly cancelling tasks, it seems like the slow operations that we might want to interrupt early are:
1. image pulling
2. rootfs extraction (is this actually slow?)
3. waiting on the container to complete

Are there any other long running steps that we should be able to abort?

Reply:
> run_until_done_or_deleted on borrow_jvm seems OK. That will let us cancel a job if JVM initialization is hung.

What is the implication of cancelling a coroutine that is awaiting the future `self._jvm_initializer_task`? This means the jvm_initializer_task will be cancelled even if it's part way through. I think we should still have `run_until_done_or_deleted`, but we should do this instead shielding the finish of the _jvm_intializer_task:

```python3
    async def borrow_jvm(self) -> JVM:
        if instance_config.worker_type() not in ('standard', 'D'):
            raise ValueError(f'JVM jobs not allowed on {instance_config.worker_type()}')
        await asyncio.shield(self._jvm_initializer_task)
        assert self._jvms
        return self._jvms.pop()
```

Reply:
We have a shield on `localize_rootfs` now so we don't have the problem with bad images anymore.

Reply:
Ah, yes, good eye. That would've cancelled the JVM initializer task. This is exactly what gives me the heebie-jeebies about `run_until_done_or_deleted`, it's so easy to miss these things. The only benefit it gives for borrow_jvm is that if the JVMs are stuck coming up we can cancel JVM jobs. But if the JVMs are stuck coming up, there's probably something very wrong with that worker.